db/external_sst_file_ingestion_job.cc

```diff
@@ -387,8 +395,16 @@ Status ExternalSstFileIngestionJob::NeedsFlush(bool* flush_needed,
 // REQUIRES: we have become the only writer by entering both write_thread_ and
 // nonmem_write_thread_
 Status ExternalSstFileIngestionJob::Run() {
-  Status status;
   SuperVersion* super_version = cfd_->GetSuperVersion();
+  // If column family is flushed after Prepare and before Run, we should have a
+  // specific state of Memtables. The mutable Memtable should be empty, and the
+  // immutable Memtable list should be empty.
+  if (flushed_before_run_ && (super_version->imm->NumNotFlushed() != 0 ||
+                              !super_version->mem->IsEmpty())) {
+    return Status::TryAgain(
+        "Inconsistent memtable state detected when flushed before run.");
+  }
+  Status status;
```

Comment:
How does write get synchronized with flush so we won't have new write arriving to a new mutable memtable while the old mutable memtable is being flushed (hence violating this check)? 

> we should have a specific state of Memtables

By "should", does that mean this check is checking an expected correct behavior? If yes, then why did we choose TryAgain() instead of Corruption() status type?

Reply:
Currently, we block writes to DB and waited for pending writes to finish on beginning file ingestion. https://github.com/facebook/rocksdb/blob/1006eddd635539466d6d99a3396de3c1df4e1993/db/db_impl/db_impl.cc#L5997-L6008

And only unblock after it finishes:
https://github.com/facebook/rocksdb/blob/1006eddd635539466d6d99a3396de3c1df4e1993/db/db_impl/db_impl.cc#L6166

The concurrent writes that are running as ingestion runs are not able to get into the new mutable memtable until that `ExitUnbatched` happens. When column family is flushed during file ingestion, it's a manual flush so everything should be persisted, the MemTables in the latest SuperVersion should have no data. 

If it's not the expected state, it definitely is worth investigating further. But maybe not necessarily as serious as a corruption, since the other `needs_flush` check is something that in theory shouldn't happen. We are returning retry there too.